### PR TITLE
Add dropping strategy

### PR DESCRIPTION
**Pull request by RissyRan**
# Description

Enable token dropping for matmul implementation

Next steps:
* work on dispatch & combine mask - make expert parallelism easy
* add load balance loss

# Test

Have a unite test to check a single layer output (become dropless if capacity factor is large enough):

* capacity factor = 2 as default (it should be 1 ideally with evenly distributed) - [link](https://paste.googleplex.com/5898720970539008#l=12) - pass
* capacity factor = 0.1 (extreme case) - [link](https://paste.googleplex.com/6179650651029504) - fail

### CODE REVIEW

**Comment by ZhiyuLi-goog** [comment]
Thank you @RissyRan for adding dropping strategy!!
I just added some nit.

**Review by tonyjohnchen — APPROVED** [review]
Thanks for adding this!

**Comment by RissyRan** [comment]
> Thank you @RissyRan for adding dropping strategy!! I just added some nit.

Thanks Zhiyu! Have you published comments?

**Comment by ZhiyuLi-goog** [comment]
> > Thank you @RissyRan for adding dropping strategy!! I just added some nit.
> 
> Thanks Zhiyu! Have you published comments?

Just published.

File: MaxText/configs/base.yml [hunk 1]
```diff
@@ -103,16 +103,18 @@ base_num_kv_heads: 16
...
 mlp_activations: ["silu", "linear"]
 dropout_rate: 0
 logits_via_embedding: False
 normalize_embedding_logits: True  # whether to normlize pre-softmax logits if logits_via_embedding is true
 logits_dot_in_fp32: True  # whether to use fp32 in logits_dense or shared_embedding dot product for stability
 
+# mixture of experts (moe)
+num_experts: 1
+num_experts_per_tok: 1
+megablox: True
```
[the remark this answers is not in the document]
Comment (RissyRan) [reply MaxText/configs/base.yml:115]: I don't think it's conflict. But the merit of megablox is dropless. So in this case, I plan to have 2 implementations:

1) dropless through megablox;
2) dropping through matmul (dropless if capacity factor is large enough)

File: MaxText/layers/linears.py [hunk 2]
```diff
@@ -410,12 +410,26 @@ def wrapper(x, logits, w0, w1, wo):
     return wrapper(inputs, gate_logits, w0_kernel, w1_kernel, wo_kernel)
 
   def dense_matmul(self, inputs, gate_logits, w0_kernel, w1_kernel, wo_kernel):
+    # shape of top_k_weights & top_k_indices: (batch, sequence, exp_per_token)
     top_k_weights, top_k_indices = jax.lax.top_k(gate_logits, self.num_experts_per_tok)
-    flattened_top_k_weights = top_k_weights.reshape(-1, self.num_experts_per_tok)
```
[the remark this answers is not in the document]
Comment (RissyRan) [reply MaxText/layers/linears.py:414]: What's the benefit of dropless in matmul when we have megablox implementation?

File: MaxText/layers/linears.py [hunk 3]
```diff
@@ -410,12 +410,26 @@ def wrapper(x, logits, w0, w1, wo):
...
+    # calculate expert_capacity = (tokens_per_batch / num_experts) * capacity_factor
+    batch_size, seq_len, _ = inputs.shape
+    tokens_per_batch = batch_size * seq_len * self.num_experts_per_tok
+    expert_capacity = int((tokens_per_batch / self.num_experts) * self.config.capacity_factor)
+    max_logging.log(f"expert_capacity is {expert_capacity}")
+
+    # calculate expert mask and drop tokens if needed
+    # shape of expert_mask: (batch, sequence, exp_per_token, num_experts)
+    expert_mask = jax.nn.one_hot(top_k_indices, num_classes=self.num_experts)
+    cumsum_expert_mask = jnp.cumsum(jnp.cumsum(expert_mask, axis=1), axis=2)
```
[the remark this answers is not in the document]
Comment (RissyRan) [reply MaxText/layers/linears.py:426]: This is a good question. The order should be kept. `jnp.cumsum` will not change the array shape. Right after this line we have `cumsum_expert_mask = cumsum_expert_mask * expert_mask`, that could map back the original one-hot. I could print an example if that helps.

File: MaxText/layers/linears.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ZhiyuLi-goog) [reply MaxText/layers/linears.py:414]: Thank you @RissyRan. I think your dropless in matmul is helpful in mlperf. Additionally, dense_matmul droppless implementation is equivalent to mixtral reference implementation and huggingface's. We can use this as a baseline for performance and numerical stability. What do you think?

[the remark this answers is not in the document]
Comment (RissyRan) [reply MaxText/layers/linears.py:414]: >dense_matmul droppless implementation is equivalent to mixtral reference implementation and huggingface's

OK, I see. I recall both of them are using for loop implementation. If we want to keep it as a reference, I can add a if/else somewhere with indicator, i.e. capacity_factor, or bool value.

File: MaxText/layers/linears.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (RissyRan) [reply MaxText/layers/linears.py:426]: Order will not be updated, we only mask weights when setting `softmax_probs`

```
    weights = jnp.zeros_like(gate_logits)
    index_update = (jnp.arange(gate_logits.shape[0])[:, None, None], jnp.arange(gate_logits.shape[1])[:, None], top_k_indices)
    weights = weights.at[index_update].set(softmax_probs)
```


No dropping example:

```
# select expert 3 & 4
top_k_indices: [[[3 4]]]

# convert to one-hot
expert_mask: [[[[0. 0. 0. 1. 0. 0. 0. 0.]
   [0. 0. 0. 0. 1. 0. 0. 0.]]]]

# those two are selected
flat_expert_mask: [[[1. 1.]]]

# softmax_probs
softmax_probs: [[[0.5019531  0.49804693]]]
```

Dropping example:

```
top_k_indices: [[[3 4]
  [4 5]
  [4 6]
  [6 4]]]

expert_mask: [[[[0. 0. 0. 1. 0. 0. 0. 0.]
   [0. 0. 0. 0. 1. 0. 0. 0.]]

  [[0. 0. 0. 0. 1. 0. 0. 0.]
   [0. 0. 0. 0. 0. 1. 0. 0.]]

  [[0. 0. 0. 0. 1. 0. 0. 0.]
   [0. 0. 0. 0. 0. 0. 1. 0.]]

  [[0. 0. 0. 0. 0. 0. 1. 0.]
   [0. 0. 0. 0. 1. 0. 0. 0.]]]]

flat_expert_mask: [[[1. 1.]
  [1. 1.]
  [1. 1.]
  [1. 0.]]]

softmax_probs: [[[0.5019531  0.49804693]
  [0.5602527  0.43974727]
  [0.5893618  0.4106382 ]
  [0.58699644 0.        ]]]

```

File: MaxText/layers/linears.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (RissyRan) [reply MaxText/layers/linears.py:414]: Updated, and no dropping by default.

File: MaxText/configs/base.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (gobbleturk) [reply MaxText/configs/base.yml:115]: an intermediate step might be expert parallelism + dropping + megablox, although an ideal solution could be dropless. E.g. if there are 64 experts and we set expert parallelism = 16 there are still 4 experts per shard, so the megablox kernel can still be useful. It seems easier to perform the expert parallelism a2a with dropping

File: MaxText/layers/linears.py [hunk 4]
```diff
@@ -410,11 +410,28 @@ def wrapper(x, logits, w0, w1, wo):
     return wrapper(inputs, gate_logits, w0_kernel, w1_kernel, wo_kernel)
 
   def dense_matmul(self, inputs, gate_logits, w0_kernel, w1_kernel, wo_kernel):
+    # shape of top_k_weights & top_k_indices: (batch, sequence, exp_per_token)
     top_k_weights, top_k_indices = jax.lax.top_k(gate_logits, self.num_experts_per_tok)
-    flattened_top_k_weights = top_k_weights.reshape(-1, self.num_experts_per_tok)
-
-    softmax_probs = jax.nn.softmax(flattened_top_k_weights.astype(jnp.float32), axis=-1).astype(self.weight_dtype)
-    softmax_probs = softmax_probs.reshape(gate_logits.shape[:-1] + (self.num_experts_per_tok,))
+    softmax_probs = jax.nn.softmax(top_k_weights.astype(jnp.float32), axis=-1).astype(self.weight_dtype)
```
Comment (gobbleturk) [inline MaxText/layers/linears.py:415]: Instead of weight_dtype shouldn't this be activation dtype?

File: MaxText/layers/linears.py [hunk 5]
```diff
@@ -410,11 +410,28 @@ def wrapper(x, logits, w0, w1, wo):
...
+
+    # token dropping
+    if self.config.capacity_factor > 0:
+      # calculate expert_capacity = (tokens_per_batch / num_experts) * capacity_factor
+      batch_size, seq_len, _ = inputs.shape
+      tokens_per_batch = batch_size * seq_len * self.num_experts_per_tok
+      expert_capacity = int((tokens_per_batch / self.num_experts) * self.config.capacity_factor)
+      max_logging.log(f"Token dropping if needed, and expert_capacity is {expert_capacity}")
```
Comment (gobbleturk) [inline MaxText/layers/linears.py:423]: nit: "Applying potential token dropping with an expert_capacity of {expert_capacity}"

File: MaxText/layers/linears.py [hunk 6]
```diff
@@ -410,11 +410,28 @@ def wrapper(x, logits, w0, w1, wo):
...
+    # shape of top_k_weights & top_k_indices: (batch, sequence, exp_per_token)
     top_k_weights, top_k_indices = jax.lax.top_k(gate_logits, self.num_experts_per_tok)
-    flattened_top_k_weights = top_k_weights.reshape(-1, self.num_experts_per_tok)
-
-    softmax_probs = jax.nn.softmax(flattened_top_k_weights.astype(jnp.float32), axis=-1).astype(self.weight_dtype)
-    softmax_probs = softmax_probs.reshape(gate_logits.shape[:-1] + (self.num_experts_per_tok,))
+    softmax_probs = jax.nn.softmax(top_k_weights.astype(jnp.float32), axis=-1).astype(self.weight_dtype)
+
+    # token dropping
+    if self.config.capacity_factor > 0:
```
Comment (gobbleturk) [inline MaxText/layers/linears.py:418]: Can we wrap the body of this in its own method - for now the method just needs to return the softmax_probs, perhaps the mask as well once we integrate this will Expert parallelism

File: MaxText/layers/linears.py [hunk 7]
```diff
@@ -410,11 +410,28 @@ def wrapper(x, logits, w0, w1, wo):
...
+      # calculate expert_capacity = (tokens_per_batch / num_experts) * capacity_factor
+      batch_size, seq_len, _ = inputs.shape
+      tokens_per_batch = batch_size * seq_len * self.num_experts_per_tok
+      expert_capacity = int((tokens_per_batch / self.num_experts) * self.config.capacity_factor)
+      max_logging.log(f"Token dropping if needed, and expert_capacity is {expert_capacity}")
+
+      # calculate expert mask and drop tokens if needed
+      # shape of expert_mask: (batch, sequence, exp_per_token, num_experts)
+      expert_mask = jax.nn.one_hot(top_k_indices, num_classes=self.num_experts)
+      cumsum_expert_mask = jnp.cumsum(jnp.cumsum(expert_mask, axis=1), axis=2)
```
Comment (gobbleturk) [inline MaxText/layers/linears.py:428]: Comments here would be helpful although it is tricky how to best explain this concisely (The code is more concise than words). I prefer the name "expert_cumusm_token_count" instead of cumsum_expert_mask for line 428, and maybe expert_token_count for line 429 (open to other ideas though). Perhaps a small example - "If expert 0 is chosen by tokens [1,3] the expert_token_count will have entries [0,1,0,2], which associates each chosen token with an increasing count"

File: MaxText/layers/linears.py (hunk 4 above)
Comment (gobbleturk) [reply MaxText/layers/linears.py:415]: oh I see you are not changing the dtype in this PR. But I think the dtype manipulation here and also here https://github.com/google/maxtext/blob/748305d5bd60d326c98d00b2274515f20b5968d9/MaxText/layers/linears.py#L338 should be changed to cast to float32 before softmax, then back to activation dtype. Perhaps the float32 cast should be configurable

File: MaxText/layers/linears.py (hunk 5 above)
Comment (RissyRan) [reply MaxText/layers/linears.py:423]: SG, will update.

File: MaxText/layers/linears.py (hunk 7 above)
Comment (RissyRan) [reply MaxText/layers/linears.py:428]: SG! Will update the variable name and make a comment there.

File: MaxText/layers/linears.py (hunk 6 above)
Comment (RissyRan) [reply MaxText/layers/linears.py:418]: Sg!

Comment (RissyRan) [reply MaxText/layers/linears.py]: Moved to generate_masks() helper method.

File: MaxText/layers/linears.py (hunk 7 above)
Comment (RissyRan) [reply MaxText/layers/linears.py]: Updated the variables, and have a small example there. Please let me know if it's unclear.

**Review by ZhiyuLi-goog — APPROVED** [review]
Thank you @RissyRan LGTM!

**Review by gobbleturk — APPROVED** [review]
Thanks Ran!
